File: server/csharp/MetaObjects.Codegen/TemplateCodegen/ScopeWalk.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MetaObjects.Codegen.Generators;
using MetaObjects.Meta;

namespace MetaObjects.Codegen.TemplateCodegen;

/// <summary>
/// The three built-in walk scopes (SP-1 §3.1): <c>perEntity</c>, <c>perPackage</c>,
/// <c>perModel</c>. Each yields the neutral data dict for its unit and names the
/// file via <see cref="OutputPattern"/>. Same vocabulary as the other ports.
/// </summary>
public static class ScopeWalk
{
    public const string PerEntity = "perEntity";
    public const string PerPackage = "perPackage";
    public const string PerModel = "perModel";

    public static readonly IReadOnlyList<string> Scopes = new[] { PerEntity, PerPackage, PerModel };

    public static Func<MetaRoot, IEnumerable<TemplateWalkResult>> ForScope(string scope, string outputPattern)
    {
        if (!Scopes.Contains(scope))
            throw new ArgumentException(
                $"unknown template scope '{scope}' (expected {string.Join(" | ", Scopes)})");
        return root =>
        {
            var objects = root.Objects();
            var concrete = objects.Where(TemplateData.IsConcrete).ToList();
            switch (scope)
            {
                case PerEntity:
                    return concrete.Select(o => new TemplateWalkResult(
                        TemplateData.Entity(o),
                        OutputPattern.Expand(outputPattern, TemplateData.BareName(o), TemplateData.PackageOf(o))))
                        .ToList();
                case PerPackage:
                    var byPkg = new SortedDictionary<string, List<MetaObject>>(StringComparer.Ordinal);
                    foreach (var o in concrete)
                    {
                        if (!byPkg.TryGetValue(TemplateData.PackageOf(o), out var list))
                            byPkg[TemplateData.PackageOf(o)] = list = new List<MetaObject>();
                        list.Add(o);
                    }
                    return byPkg.Select(kv => new TemplateWalkResult(
                        TemplateData.Package(kv.Key, kv.Value),
                        OutputPattern.Expand(outputPattern, null, kv.Key))).ToList();
                case PerModel:
                    return new[]
                    {
                        new TemplateWalkResult(
                            TemplateData.Model(objects),
                            OutputPattern.Expand(outputPattern, null, null)),
                    };
                default:
                    throw new ArgumentException(
                        $"unknown template scope '{scope}' (expected {string.Join(" | ", Scopes)})");
            }
        };
    }
}

File: server/java/codegen-base/src/main/java/com/metaobjects/generator/template/ScopeWalk.java
package com.metaobjects.generator.template;

import com.metaobjects.object.MetaObject;
import com.metaobjects.render.templategen.TemplateWalkResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Function;

/**
 * The three built-in walk scopes (SP-1 §3.1): {@code perEntity}, {@code perPackage},
 * {@code perModel}. Each yields the neutral data dict for its unit and names the
 * file via {@link OutputPattern}. Same vocabulary as the TS engine helpers.
 */
public final class ScopeWalk {

    private ScopeWalk() {}

    public static final String PER_ENTITY = "perEntity";
    public static final String PER_PACKAGE = "perPackage";
    public static final String PER_MODEL = "perModel";

    public static Function<List<MetaObject>, List<TemplateWalkResult>> forScope(
            String scope, String outputPattern) {
        return objects -> {
            List<MetaObject> concrete = new ArrayList<>();
            for (MetaObject o : objects) {
                if (TemplateData.isConcrete(o)) concrete.add(o);
            }
            List<TemplateWalkResult> out = new ArrayList<>();
            switch (scope) {
                case PER_ENTITY:
                    for (MetaObject o : concrete) {
                        out.add(new TemplateWalkResult(
                            TemplateData.entity(o),
                            OutputPattern.expand(outputPattern, TemplateData.bareName(o), TemplateData.packageOf(o))));
                    }
                    return out;
                case PER_PACKAGE: {
                    Map<String, List<MetaObject>> byPkg = new TreeMap<>();
                    for (MetaObject o : concrete) {
                        byPkg.computeIfAbsent(TemplateData.packageOf(o), k -> new ArrayList<>()).add(o);
                    }
                    for (Map.Entry<String, List<MetaObject>> e : byPkg.entrySet()) {
                        out.add(new TemplateWalkResult(
                            TemplateData.pkg(e.getKey(), e.getValue()),
                            OutputPattern.expand(outputPattern, null, e.getKey())));
                    }
                    return out;
                }
                case PER_MODEL:
                    out.add(new TemplateWalkResult(
                        TemplateData.model(objects),
                        OutputPattern.expand(outputPattern, null, null)));
                    return out;
                default:
                    throw new IllegalArgumentException(
                        "unknown template scope '" + scope + "' (expected "
                        + PER_ENTITY + " | " + PER_PACKAGE + " | " + PER_MODEL + ")");
            }
        };
    }
}
